x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunner.java

```diff
@@ -353,106 +346,24 @@ public static ClusterState setPolicyForIndexes(final String newPolicyName, final
-
-    private static boolean isActionChanged(StepKey stepKey, LifecyclePolicy currentPolicy, LifecyclePolicy newPolicy) {
-        LifecycleAction currentAction = getActionFromPolicy(currentPolicy, stepKey.getPhase(), stepKey.getAction());
-        LifecycleAction newAction = getActionFromPolicy(newPolicy, stepKey.getPhase(), stepKey.getAction());
-        if (newAction == null) {
-            return true;
-        } else {
-            return currentAction.equals(newAction) == false;
-        }
-    }
-
-    private static LifecycleAction getActionFromPolicy(LifecyclePolicy policy, String phaseName, String actionName) {
-        Phase phase = policy.getPhases().get(phaseName);
-        if (phase != null) {
-            return phase.getActions().get(actionName);
-        } else {
-            return null;
-        }
-    }
 
-    /**
-     * Returns <code>true</code> if the provided policy is allowed to be updated
-     * given the current {@link ClusterState}. In practice this method checks
-     * that all the indexes using the provided <code>policyName</code> is in a
-     * state where it is able to deal with the policy being updated to
-     * <code>newPolicy</code>. If any of these indexes is not in a state wheree
-     * it can deal with the update the method will return <code>false</code>.
-     *
-     * @param policyName
-     *            the name of the policy being updated
-     * @param newPolicy
-     *            the new version of the {@link LifecyclePolicy}
-     * @param currentState
-     *            the current {@link ClusterState}
-     */
-    public static boolean canUpdatePolicy(String policyName, LifecyclePolicy newPolicy, ClusterState currentState) {
-        Map<String, LifecyclePolicy> policiesMap = ((IndexLifecycleMetadata) currentState.metaData().custom(IndexLifecycleMetadata.TYPE))
-                .getPolicies();
-        for (ObjectCursor<IndexMetaData> cursor : currentState.getMetaData().indices().values()) {
-            IndexMetaData idxMetadata = cursor.value;
-            Settings idxSettings = idxMetadata.getSettings();
-            String currentPolicyName = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(idxSettings);
-            LifecyclePolicy currentPolicy = null;
-            if (Strings.hasLength(currentPolicyName)) {
-                currentPolicy = policiesMap.get(currentPolicyName);
-            }
-            if (policyName.equals(currentPolicyName)) {
-                StepKey currentStepKey = IndexLifecycleRunner.getCurrentStepKey(idxSettings);
-                if (canSetPolicy(currentStepKey, currentPolicy, newPolicy) == false) {
-                    return false;
-                }
+        Settings.Builder newSettings = Settings.builder().put(idxSettings);
+        if (currentStepKey != null) {
+            // Check if current step exists in new policy and if not move to
+            // next available step
+            StepKey nextValidStepKey = newPolicy.getNextValidStep(currentStepKey);
```

Comment:
I thinks the getNextValidStep() stuff can go to no?

Reply:
Can I do that as a followup? I think this change ended up larger than I hoped due to having to change a lot of tests

Reply:
Sure.